opm/output/eclipse/EclipseWriter.cpp

```diff
@@ -104,171 +77,98 @@ void restrictAndReorderToActiveCells(std::vector<double> &data,
-    Keyword(const std::string& name,
-            const std::vector<int>& data)
-        : ertHandle_(0)
-    { set(name, data); }
-
-    Keyword(const std::string& name,
-            const std::vector<const char*>& data)
-        : ertHandle_(0)
-    {set(name, data); }
-
-    ~Keyword()
+    template< typename U >
+    Keyword( const std::string& name, const std::vector< U >& data ) :
+        handle( ecl_kw_alloc( name.c_str(), data.size(), ert_data_type< T >::type ) )
     {
-        if (ertHandle_)
-            ecl_kw_free(ertHandle_);
-    }
-
-    template <class DataElementType>
-    void set(const std::string name, const std::vector<DataElementType>& data)
-    {
-
-        if(ertHandle_) {
-            ecl_kw_free(ertHandle_);
-        }
+        T* target = static_cast< T* >( ecl_kw_get_ptr( this->handle.get() ) );
 
-
-        ertHandle_ = ecl_kw_alloc(name.c_str(),
-                                  data.size(),
-                                  ertType_());
-
-        // number of elements to take
-        const int numEntries = data.size();
-
-        // fill it with values
-
-        T* target = static_cast<T*>(ecl_kw_get_ptr(ertHandle()));
-        for (int i = 0; i < numEntries; ++i) {
-            target[i] = static_cast<T>(data[i]);
-        }
+        for( size_t i = 0; i < data.size(); ++i )
+            target[ i ] = T( data[ i ] );
     }
 
-    void set(const std::string name, const std::vector<const char *>& data)
+    Keyword( const std::string& name, const std::vector< const char* >& data ) :
+        handle( ecl_kw_alloc( name.c_str(), data.size(), ECL_CHAR_TYPE ) )
     {
-      if(ertHandle_) {
-          ecl_kw_free(ertHandle_);
-      }
-
-
-      ertHandle_ = ecl_kw_alloc(name.c_str(),
-                                data.size(),
-                                ertType_());
-
-      // number of elements to take
-      const int numEntries = data.size();
-      for (int i = 0; i < numEntries; ++i) {
-          ecl_kw_iset_char_ptr( ertHandle_, i, data[i]);
-     }
-
+        for( size_t i = 0; i < data.size(); ++i )
+          ecl_kw_iset_char_ptr( this->handle.get(), i, data[ i ] );
     }
 
-    ecl_kw_type *ertHandle() const
-    { return ertHandle_; }
+    ecl_kw_type *ertHandle() { return this->handle.get(); }
+    const ecl_kw_type *ertHandle() const { return this->handle.get(); }
 
 private:
-    static ecl_type_enum ertType_()
-    {
-        if (std::is_same<T, float>::value)
-        { return ECL_FLOAT_TYPE; }
-        if (std::is_same<T, double>::value)
-        { return ECL_DOUBLE_TYPE; }
-        if (std::is_same<T, int>::value)
-        { return ECL_INT_TYPE; }
-        if (std::is_same<T, const char *>::value)
-        { return ECL_CHAR_TYPE; }
-
-
-        OPM_THROW(std::logic_error,
-                  "Unhandled type for data elements in EclipseWriterDetails::Keyword");
-    }
-
-    ecl_kw_type *ertHandle_;
+    ERT::ert_unique_ptr< ecl_kw_type, ecl_kw_free > handle;
```

Comment:
Again - here I hope you can use: https://github.com/Ensembles/ert/blob/master/devel/libeclxx/src/EclKW.cpp - if not we should improve the ert class.

If it does not make sense to use the ert class please find a better name than `handle`[1] - handle to what??

[1]: I know the old used that name - and it always annoyed me.